text/3323-restrictions.md

```diff
@@ -0,0 +1,466 @@
+what is required for backwards compatibility with existing code.
+
+`pub(mod)` is equivalent to `pub(self)`; likewise for `impl(mod)` and `mut(mod)`.
+
+When an `ImplRestriction` is present, implementations of the associated trait are only permitted
+within the designated path. Any implementation of the trait outside this scope is a compile error.
+When a `MutRestriction` is present, mutable uses of the associated field are only permitted within
+the designated path. Any mutable use of the field outside the scope is a compile error. Further, a
+`struct`, `union`, or `enum` variant containing fields with an associated `MutRestriction` may not
+be constructed with `struct` expressions unless all fields are unrestricted in the present scope.
+This is the case even if the field is not directly declared, such as when functional record updates
+are used.
+
+## "Mutable use" in the compiler
+
+The concept of a "mutable use" [already exists][mutating use method] within the compiler. This
+catches all situations that are relevant here, including `ptr::addr_of_mut!`, `&mut`, and direct
+assignment to a field, while excluding interior mutability. As such, formal semantics of what
+constitutes a "mutable use" are not stated here.
+
+[mutating use method]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/visit/enum.PlaceContext.html#method.is_mutating_use
+
+## Interaction with `trait` aliases
+
+Trait aliases cannot be implemented. As such, there is no concern about compatibility between the
+`impl` restriction and `trait` aliases.
+
+# Drawbacks
+
+- Additional syntax for macros to handle
+
+# Alternatives
+
+- `impl` and `mut` restrictions could be attributes, similar to `#[non_exhaustive]`.
+
+# Prior art
+
+- The [`readonly` crate] simulates immutable fields outside of the defining module. Types with this
+  attribute cannot define `Deref`, which can be limiting. Additionally, it applies to all fields and
+  within the defining crate. The advantages of native read-only fields relating to borrow checking
+  also do not apply when using this crate.
+- The ability to restrict implementations of a trait can be simulated by a public trait in a private
+  module. This has the disadvantage that the trait is no longer nameable by external users,
+  preventing its use as a generic bound. Current diagnostics, while technically correct, are
+  unhelpful to downstream users.
+- Various other languages have read-only fields, including C++, C#, Java, TypeScript, Kotlin, and
+  Swift.
+
+# Unresolved questions
+
+- Should unrestricted "restrictions" be allowed to be explicitly stated, or should this be banned?
+  If allowed, should it be a warning?
+  - Both `impl` and `mut` make no sense to have unrestricted, as that is the default behavior. An
+    "unnecessary" restriction lint that is warn-by-default would be ideal. This would allow macros
+    to emit more flexible syntax. This warning could also be used for `pub(self)`.
```

Comment:
I see; macros do seem like the *one* case where this might potentially be useful.

Reply:
Though for macros it might be more useful to allow `impl()` and `mut()` maybe?